### ISSUE TITLE
Requests with Large POST Body Immediately Goes into Delay Cancellation Right After the Server Received the Request

### ISSUE THREAD

**Issue by raphaelNguyen** [playframework/playframework#12351]
### Play Version
3.0.0 and 2.9.0

### API
Both Scala and Java

### Operating System
Windows

### JDK
17.0.9

### Expected Behavior
Server do not put the connection into delayed cancellation (triggering `pekko`/`akka` `linger-timeout`) right after the request is received and routed to the play `Action`. Connection is kept open until the server has responded.

### Actual Behavior
When the client has `KeepAlive` turned off and the request body is sufficiently large (so `pekko-http` received it as 2 chunks), the server put the connection into delayed cancellation right after the request is received and routed to play `Action` with the debug message below. The same request (with `KeepAlive` off) but having a smaller request body does not have its connection put into delayed cancellation by the server.

```
[DEBUG] o.a.p.h.i.u.StreamUtils$DelayCancellationStage$$anon$11: Delaying cancellation for 1 minute
```

#### Further Info
The issue is not reproducible on my old play 2.8 environment, but is reproducible when upgrading to both 2.9 and 3.0. Please see table below.

| `play-framework` version  | `akka`/`pekko` version | `akka-http`/`pekko-http` version | Issue reproducible |
|-----------------------------|--------------------------|------------------------------------|----------------------|
| `2.8.20`                              | `2.6.21`                         | `10.2.9` (*override)                      | NO                         |
| `2.9.0`                                | `2.6.21`                         | `10.2.10`                                      | YES                         |
| `2.9.0`                                | `2.6.21`                         | `10.2.9` (**override)                     | YES                         |
| `3.0.0`                                | `1.0.1`                           | `1.0.0`                                          | YES                         |

*Note: Prior to upgrading, when still using `play-framework:2.8.x` with `akka-http:10.1.x` that came with it, I also observed an issue with the same symptoms. At this time, we noticed that upgrading to `akka-http:10.2.1` fixed this issue and so I've been overriding `akka-http` to a version > `10.2.1` while using `play-framework:2.8.x`.

**Note: After reproducing the issue on `play-framework:2.9.0`, I went one step futher and forced the version of `akka-http` back to `10.2.9`, which we were using previously and was still able to reproduce the issue. This and some discussion with `pekko-http` team on https://github.com/apache/incubator-pekko-http/issues/460 (about this same issue) make me inclined to think the way `play-framework` uses `akka`/`pekko` might be causing this behaviour.

For logs, please see the next section

### Reproducible Test Case

Minimum Reproducible Project - https://github.com/integrated-application-development/play-premature-close-connection

This is a minimal play application that has only 1 `/` route. In the corresponding activity, it logs the received `POST` body and then sleeps for 65 seconds before returning a minimal `Welcome to Play` HTML page.

I run the server with `-Dhttp.port=<server-port>` and `-Dhttps.port=disabled`. Then I hit this application with the following requests from powershell.

```powershell
Invoke-WebRequest -Method POST -Uri http://localhost:<server-port>/ -Body $("0"*130000) -ContentType "text/plain" -DisableKeepAlive;
Invoke-WebRequest -Method POST -Uri http://localhost:<server-port>/ -Body $("0"*140000) -ContentType "text/plain" -DisableKeepAlive;
```

Edit by @mkurz, added the equivalent `curl` commands:
```sh
curl -X POST http://localhost:<server-port>/ -d "$(printf '0%.0s' {1..130000})" -H "Content-Type: text/plain" -H "Connection: close"
curl -X POST http://localhost:<server-port>/ -d "$(printf '0%.0s' {1..140000})" -H "Content-Type: text/plain" -H "Connection: close"
```

The first request returns correctly after 65 seconds (sleep time) and the second request has the connection unexpected closed after roughly 60 seconds (default `pekko.http.server.linger-timeout` is 1 minute).

**Note** Sometimes, depending on the connection speed, the POST body size threshold to trigger the issue might change.

The following log is obtained

```
2024-02-01 11:56:23 DEBUG org.apache.pekko.io.TcpListener  New connection accepted
2024-02-01 11:56:23 TRACE play.core.server.PekkoHttpServer  Http request received by pekko-http: HttpRequest(HttpMethod(POST),http://localhost:<server-port>/,List(Timeout-Access: <function1>, Remote-Address, Raw-Request-URI: /, Tls-Session-Info: Session(1706748963006|SSL_NULL_WITH_NULL_NULL), User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; en-GB) WindowsPowerShell/5.1.19041.3930, Host, Connection: Close),HttpEntity.Strict(text/plain,130000 bytes total),HttpProtocol(HTTP/1.1))
2024-02-01 11:56:23 TRACE p.c.r.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5  Not deferring body parsing for request, will parse now: POST /
2024-02-01 11:56:23 TRACE play.api.mvc.BodyParser  Invoking action with request: POST /
2024-02-01 11:56:23 DEBUG play.mvc.Action  ### Start of action order
2024-02-01 11:56:23 DEBUG play.mvc.Action  1. play.http.DefaultActionCreator$1()
2024-02-01 11:56:23 DEBUG play.mvc.Action  2. play.core.j.JavaAction$$anon$1()
2024-02-01 11:56:23 DEBUG play.mvc.Action  ### End of action order
2024-02-01 11:56:23 TRACE play.api.mvc.BodyParser  Not parsing body, it's a WebSocket or it was parsed before already for request: POST /
2024-02-01 11:56:23 INFO  controllers.HomeController  Body received: 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000[...] (130000 chars long)
2024-02-01 11:57:29 DEBUG org.apache.pekko.io.TcpListener  New connection accepted
2024-02-01 11:57:29 TRACE play.core.server.PekkoHttpServer  Http request received by pekko-http: HttpRequest(HttpMethod(POST),http://localhost:<server-port>/,List(Timeout-Access: <function1>, Remote-Address, Raw-Request-URI: /, Tls-Session-Info: Session(1706748963006|SSL_NULL_WITH_NULL_NULL), User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; en-GB) WindowsPowerShell/5.1.19041.3930, Host, Connection: Close),HttpEntity.Default(text/plain,140000 bytes total),HttpProtocol(HTTP/1.1))
2024-02-01 11:57:29 TRACE p.c.r.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5  Not deferring body parsing for request, will parse now: POST /
2024-02-01 11:57:29 TRACE play.api.mvc.BodyParser  Invoking action with request: POST /
2024-02-01 11:57:29 DEBUG play.mvc.Action  ### Start of action order
2024-02-01 11:57:29 DEBUG play.mvc.Action  1. play.http.DefaultActionCreator$1()
2024-02-01 11:57:29 DEBUG play.mvc.Action  2. play.core.j.JavaAction$$anon$1()
2024-02-01 11:57:29 DEBUG play.mvc.Action  ### End of action order
2024-02-01 11:57:29 TRACE play.api.mvc.BodyParser  Not parsing body, it's a WebSocket or it was parsed before already for request: POST /
2024-02-01 11:57:29 INFO  controllers.HomeController  Body received: 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000[...] (140000 chars long)
2024-02-01 11:57:29 DEBUG o.a.p.h.i.u.StreamUtils$DelayCancellationStage$$anon$11  Delaying cancellation for 1 minute
2024-02-01 11:58:29 DEBUG o.a.p.h.i.u.StreamUtils$DelayCancellationStage$$anon$11  Stage was canceled after delay of 1 minute
2024-02-01 11:58:30 DEBUG o.a.p.s.i.i.TcpConnectionStage$TcpStreamLogic  Closing connection from 0:0:0:0:0:0:0:1:<client-port> because downstream cancelled stream without failure
```

In addition to this debug log, I also put in a few break points on [pekko's GraphInterpreter.scala](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala) to log some debug information at the following lines to monitor the way the request was being received.
* [L516-L518](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L516-L518)
* [Cancellation cause at L521](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L521)
* [L530-L532](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L530-L532)
* [L550-L552](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L550-L552)
* [L561-L563](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L561-L563)
* [L638](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L638)
* [L653](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L653)
* [L672](https://github.com/apache/incubator-pekko/blob/a3f48833c97501a0193f4135f99c5f6ea04a03dd/stream/src/main/scala/org/apache/pekko/stream/impl/fusing/GraphInterpreter.scala#L672)

Please find those extended logs here if they can be of any help (these are the same 2 requests as seen above)
* No delay cancellation - [play-trace-and-communication-breakpoints--130000-body.log](https://github.com/playframework/playframework/files/14135915/play-trace-and-communication-breakpoints--130000-body.log)
* Delay cancellation - [play-trace-and-communication-breakpoints--140000-body.log](https://github.com/playframework/playframework/files/14135905/play-trace-and-communication-breakpoints--140000-body.log)

Looking at these logs, we can see that, in the error case (140000 characters), the request is received as 2 chunks while, in the normal case, it is received as 1 chunk. https://github.com/apache/incubator-pekko-http/issues/460 has some comments from `pekko-http` regarding this difference.

Thank you very much for your time and if you require more information, please do not hesitate to ask.

**Comment by raphaelNguyen** [comment]
Can I ask if there might be some progress on this? Thank you for your time.

**Comment by mkurz** [comment]
@raphaelNguyen Thanks for your detailed report and the reproducer. I was able to reproduce locally as well. I can confirm this problem occurs with Play 2.9 and 3.0.

At first I was a bit confused, because I also tried Play 2.8.21 with akka-http upgraded to 10.2.10, the same version that Play 2.9.1 uses: With 2.8 it worked, but with 2.9 it failed (like you said)... I took me quite a while to figure out what the problem is, like what the difference in Play's akka http server backend is in 2.8 vs. 2.9...

Turns out the problem only occurs when [HTTP pipelining](https://en.wikipedia.org/wiki/HTTP_pipelining) ([RFC](https://datatracker.ietf.org/doc/html/rfc7230#section-6.3.2)) is enabled... which actually was the default in akka-http 10.1.x, but was disabled in 10.2.x (https://github.com/akka/akka-http/pull/3321) (and therefore also Pekko v1).

So... despite akka-http disabling it in 10.2.x, Play did not:
- #11918

this was done mainly because benchmarks did suffer from disabling it and I thought we can just keep the setting enabled, since it seemed it does not hurt and akka-http had it enabled since years... I didn't know this would cause problems. It seems that meanwhile akka's internal code changed in regards of http pipelining so that problem like yours can surface....
Given that basically HTTP pipelining is more or less used for benchmarks, and not supported anymore in any major browser, nor `curl` (see [here](https://daniel.haxx.se/blog/2019/04/06/curl-says-bye-bye-to-pipelining/)) I think we are safe to  just disable it as well, given that akka-http did so and they say it's an obscure feature which isn't tested well (see the comments the above issue). Also Netty does not even support it.

Anyway, can you try to set:
```
PlayKeys.devSettings += "play.server.akka.pipelining-limit" -> "1" // Play 2.9
PlayKeys.devSettings += "play.server.pekko.pipelining-limit" -> "1" // Play 3.0
```
in your `build.sbt` and let me know if this fixes your problem? Those lines disable http pipelining.

You also have to put it into your `conf/application.conf` file, because `PlayKeys.devSettings` are only for dev mode:
```
# Play 2.9:
play.server.akka.pipelining-limit = 1
# Play 3.0:
play.server.pekko.pipelining-limit = 1
```

If you can confirm this fixes the problem, we will disable http pipelining with next bug fix release by default.

**Comment by jrudolph** [comment]
Thanks for looking into it, @mkurz. In general, we had HTTP pipelining indeed enabled for a long time but (as you mentioned) aside from benchmarks it is not recommended to be used in practice and there is little actual usage from any kind of common client libraries in default settings. One reason it is not recommended, is that you get head-of-line blocking and it is much more likely to overload a server with requests that might not even be served in the end (if the connection crashes for whatever reason).

So, in general pipelining support is there but it is not tested to the same degree that regular HTTP/1.1 support is and that's why disabled it for 10.2.

Regarding benchmarking, if you want to be fast in the pipelining benchmarks, you need to expose the pipelining setting and only turn it on for the benchmark. Pipelining can be beneficial in certain situations and imo it's fine to show those off in a benchmark but it's not thought to be turned on by default.

**Comment by jrudolph** [comment]
Btw. I'm not 100% sure that this actually is *only* caused by pipelining.

In my investigations, there was some indications that code in Play uses the `head` stream operator to ignore a part of the request entity. Since `head` will cancel the request entity stream, using `head` is not safe to do if you want to return a response (or keep using the connection). This has been under-specified in Akka HTTP (see e.g. https://github.com/akka/akka-http/issues/3458). The rationale is that you cannot cancel an incoming request entity in HTTP/1.1 on the protocol level, so taking the cancellation signal seriously means we have to close connection immediately. I cannot remember exactly any more but I assume that for implementation reasons, returning a response might still work (probably because of the cancellation delay) but I would consider this accidental. This would also explain why pipelining is more certainly affected: eventually the connection will be closed after a request entity cancellation and if you already have more requests in-flight on that connection, all these will be affected.

The real fix in play should be to make sure that `head`, `headOption` and all other stream cancelling operators from request entity processing (BodyPartParser) do not cancel the request entity stream if you still want to return a response. You could replace them with another operator that just slurps in and ignores the rest of the stream e.g. a general purpose GraphStage that replaces a downstream cancellation signal with ignoring the rest of the stream and put it somewhere in front of the `head` in the request entity processing.

Arguably, a behavior like this could also be provided in Pekko HTTP directly under a setting (but currently isn't).

**Comment by mkurz** [comment]
@jrudolph Thanks. I might investigate when I find time (I did so right now quickly... but not into depth). Given the issue is resolved with disabling http pipelining I just leave things as they are now. However, if you want to come up with a pull request where we can discuss things more detailed, I am very happy if you do so.

**Comment by raphaelNguyen** [comment]
Thank you for taking the time to investigate this issue, @mkurz, @jrudolph. I can confirm that turning off pipelining looks to have solved the issue I was experiencing. I will be turning it off manually until play release the next version. Apologies for the delay in getting back to you.

**Comment by sedhha** [comment]
So if I am having this connection drop issue I am supposed to disable pipelining right by doing:
```
play.server.akka.pipelining-limit = 0
```

**Comment by mkurz** [comment]
@sedhha Which Play version are you using? This should be fixed in Play 2.9.2/3.0.2 already.

**Comment by sedhha** [comment]
I am using play 3.0.3. Actually the API works fine however while running automation tests from java, I get 503 connection terminated. Could it be because of `HTTP` version or something?
